TicTacToe/TicTacToe_hardcore.py

```diff
@@ -0,0 +1,183 @@
+        check.append(board[index][reverse_index])
+    if all_same(check):
+        return True
+
+    return False
+
+# checks if the game is a tie
+def tie():
+    for row in board:
+        for spot in row:
+            if spot == " ":
+                return False
+
+    return True
+
+# main game loop
+playing = True
+while playing:
+    # asks player what board size they want
+    # if choice is less than 1 or causes an error, it repeats
+    board = []
+    getting_board_size = True
+    while getting_board_size:
+        try:
+            board_size = int(input("What board size? "))
+
+            if board_size <= 0:
+                raise Exception()
+            else:
+                getting_board_size = False
+        except:
+            print("Invalid choice!")
+
+    # sets the board's number of rows and columns equal to the chosen board size
+    for row_size in range(board_size):
+        row = []
+        for column_size in range(board_size):
+            row.append(" ")
+        board.append(row)
+
+    current_player = player_X
+
+    # loop that repeats while game is ongoing
+    game_ongoing = True
+    while game_ongoing:
+        # displays board
+        print_board()
+
+        # displays whose turn it is
+        print(f"\n{current_player}'s turn!")
+
+        # asks player where to place their mark, 
+        # if choice is occupied, doesn't exist, or causes an error, it repeats
+        played = False
+        while not played:
+            try:
+                column = int(input("Which column? "))
+                row = int(input("Which row? "))
+
+                if board[row][column] == " ":
+                    board[row][column] = current_player
+                    played = True
+                else:
+                    raise Exception()
+            except:
+                print("Invalid choice!")
+
+        # checks if someone wins or there's a tie
+        # if so, displays the winner, increments their win count, and ends ongoing game loop
+        # if not, gives turn to next player and repeats ongoing game loop
+        if win():
+            print_board()
+            print(f"\n{current_player} wins!")
+            wins[current_player] += 1
+            game_ongoing = False
+        elif tie():
+            print_board()
+            print(f"\nTie!")
+            game_ongoing = False
+        else:
+            if current_player == player_X:
+                current_player = player_O
+            else:
+                current_player = player_X 
+
+    # displays the win counter for each player
+    print(f"X's wins: {wins[player_X]}")
+    print(f"O's wins: {wins[player_O]}")
+
+    # asks player if they want to play another game
+    # if yes, repeats the main game loop
+    # if no, ends main game loop
+    getting_again_choice = True
+    while getting_again_choice:
+        again_choice = input("Play another game? (y/n) ")
```

Comment:
Small thing but in case they (for whatever reason) decide to type in uppercase, this might be better:
`again_choice = input("Play another game? (y/n) ").lower()`